### ISSUE TITLE
clone & reset handling different than feathers-vuex

### ISSUE THREAD

**Issue by fratzinger** [marshallswain/feathers-pinia#30]
In `feathers-vuex` cloning a clone isn't possible. `item.clone().clone()` throws an error: `'You cannot clone a copy'`.

```ts
// see https://github.com/feathersjs-ecosystem/feathers-vuex/blob/8643bd7b0cd045fcfd1e407143c3221ae58b7486/src/service-module/make-base-model.ts#L305-L313
if (this.__isClone) {
  throw new Error('You cannot clone a copy')
}
```

In `feathers-pinia` cloning a clone is totally possible. You already have a test for it. See:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/tests/model-instance-clone-and-commit.test.ts#L39-L47

Don't know, if it's intended, though. In `feathers-pinia` there is `item.reset()` as in `feathers-vuex`, but it calls `store.resetCopy()`, which is not existent. See:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/base-model.ts#L131-L135

If `clone.clone()` is intended to work, that's a breaking change from `feathers-vuex` worth mentioning in the docs!

Also constructor adds `__isClone` differently compared to `feathers-vuex`. In `feathers-vuex` it only gets added when it should be a clone:
```ts
// see https://github.com/feathersjs-ecosystem/feathers-vuex/blob/8643bd7b0cd045fcfd1e407143c3221ae58b7486/src/service-module/make-base-model.ts#L151-L157
// Mark as a clone
if (options.clone) {
  Object.defineProperty(this, '__isClone', {
    value: true,
    enumerable: false
  })
}
```
In `feathers-pinia` it always gets added as `true` or `false`:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/base-model.ts#L24

You also make it explicitely defined in (see exclamation mark):
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/base-model.ts#L18

But on `clone.commit()` you remove it. You have a test for that:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/tests/model-instance-clone-and-commit.test.ts#L27-L37

That's because `fast-copy` does not add the `enumerable: false` property of `__isClone`. Whats the intended way?
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/make-actions.ts#L343

Todos:
- [x] say if `clone.clone()` should work or not @marshallswain 
- [x] say if `clone.__isClone` should always return a boolean or if it is not existent @marshallswain 
- [x] get `clone.reset()` running again, add old `resetCopy` mutation as `resetClone` action @fratzinger 
- [x] add `resetClone` renaming to the 'What's Different from Feathers-Vuex' section
- [x] add a test for `clone.reset()` @fratzinger 
- [ ] If `clone.clone()` is intended to work, add it to the 'What's Different from Feathers-Vuex' section @fratzinger 
- [ ] ~If `clone.__isClone` should always return a boolean, add `Object.defineProperty` to `store.commit`; change the test of `commit()` accordingly @fratzinger~

**Comment by marshallswain** [comment]
For feathers-pinia, I reevaluated the logic, yes.  Instead of throwing
an error on cloning a clone, it should work the same as reset.  You can
only have one copy of a clone, really.  So if you call clone.clone(),
any existing clone will be reset.

The other property you asked about should be Boolean | undefined

On January 19, 2022, GitHub ***@***.***> wrote:
> In feathers-vuex cloning a clone isn't possible. item.clone().clone()
> throws an error: 'You cannot clone a copy'.
>
> // see https://github.com/feathersjs-ecosystem/feathers-
> vuex/blob/8643bd7b0cd045fcfd1e407143c3221ae58b7486/src/service-
> module/make-base-model.ts#L305-L313 if (this.__isClone) { throw new
> Error('You cannot clone a copy') }
> In feathers-pinia cloning a clone is totally possible. You already
> have a test for it. See:
>  <https://github.com/marshallswain/feathers-
> pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/tests/model-
> instance-clone-and-commit.test.ts#L39-L47>
>
> Don't know, if it's intended, though. In feathers-pinia there is
> item.reset() as in feathers-vuex, but it calls store.resetCopy(),
> which is not existent. See:
>  <https://github.com/marshallswain/feathers-
> pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-
> store/base-model.ts#L131-L135>
>
> If clone.clone() is intended to work, that's a breaking change from
> feathers-vuex worth mentioning in the docs!
>
> Also constructor adds __isClone differently compared to feathers-vuex.
> In feathers-vuex it only gets added when it should be a clone:
>
> // see https://github.com/feathersjs-ecosystem/feathers-
> vuex/blob/8643bd7b0cd045fcfd1e407143c3221ae58b7486/src/service-
> module/make-base-model.ts#L151-L157 // Mark as a clone if
> (options.clone) { Object.defineProperty(this, '__isClone', { value:
> true, enumerable: false }) }
> In feathers-pinia it always gets added as true or false:
>  <https://github.com/marshallswain/feathers-
> pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-
> store/base-model.ts#L24>
>
> You also make it explicitely defined in (see exclamation mark):
>  <https://github.com/marshallswain/feathers-
> pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-
> store/base-model.ts#L18>
>
> But on clone.commit() you remove it. You have a test for that:
>  <https://github.com/marshallswain/feathers-
> pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/tests/model-
> instance-clone-and-commit.test.ts#L27-L37>
>
> That's because fast-copy does not add the enumerable: false property
> of __isClone. Whats the intended way?
>  <https://github.com/marshallswain/feathers-
> pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-
> store/make-actions.ts#L343>
>
> Todos:
>
>  *   say if clone.clone() should work or not @marshallswain
>     <https://github.com/marshallswain>
>  *   say if clone.__isClone should always return a boolean or if it is
>     not existent @marshallswain <https://github.com/marshallswain>
>  *   get clone.reset() running again, add old resetCopy mutation as
>     resetClone action @fratzinger <https://github.com/fratzinger>
>  *   add resetClone renaming to the 'What's Different from Feathers-
>     Vuex' section
>  *   add a test for clone.reset() @fratzinger
>     <https://github.com/fratzinger>
>  *   If clone.clone() is intended to work, add it to the 'What's
>     Different from Feathers-Vuex' section @fratzinger
>     <https://github.com/fratzinger>
>  *   If clone.__isClone should always return a boolean, add
>     Object.defineProperty to store.commit; change the test of commit()
>     accordingly @fratzinger <https://github.com/fratzinger>
>

—
Reply to this email directly, view it on GitHub
<https://github.com/marshallswain/feathers-pinia/issues/30>, or
unsubscribe <https://github.com/notifications/unsubscribe-
auth/AAA7OWJDJUXQAEA5I6OFGJTUWZ6P7ANCNFSM5MJLYJ3A>.
You are receiving this because you were mentioned.Message ID:
***@***.***>

**Comment by marshallswain** [comment]
clone.__isClone can be Boolean or undefined

**Comment by marshallswain** [comment]
This is fixed in `feathers-pinia@0.31.0`.
